go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -75,30 +75,48 @@ type serviceConfig struct {
-				cookie, npw.ofportPhys, flowProtocol, nwDst, externalIPOrLBIngressIP, svcPort, HostNodePortCTZone),
-			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, %s=%s, tp_src=%d, actions=ct(zone=%d,table=7)",
-				cookie, flowProtocol, nwSrc, externalIPOrLBIngressIP, svcPort, HostNodePortCTZone))
+		// table 0, This rule matches on all traffic with dst ip == LoadbalancerIP / externalIP, DNAT's it to the correct NodeIP
+		if strings.Contains(flowProtocol, "6") {
+			// If ipv6 make sure to choose the ipv6 node address for rule
+			nodeportFlows = append(nodeportFlows,
+				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, %s=%s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=[%s]:%s),table=6)",
+					cookie, npw.ofportPhys, flowProtocol, nwDst, externalIPOrLBIngressIP, svcPort.Port, HostNodePortCTZone, npw.gatewayIPv6, svcPort.TargetPort.String()))
+		} else {
+			nodeportFlows = append(nodeportFlows,
+				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, %s=%s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=%s:%s),table=6)",
+					cookie, npw.ofportPhys, flowProtocol, nwDst, externalIPOrLBIngressIP, svcPort.Port, HostNodePortCTZone, npw.gatewayIPv4, svcPort.TargetPort.String()))
+		}
+		nodeportFlows = append(nodeportFlows,
+			// table 0, Matches on return traffic, i.e traffic coming from the host networked pod's port, and unDNATs
+			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, tp_src=%s, actions=ct(commit,zone=%d nat,table=7)",
+				cookie, flowProtocol, svcPort.TargetPort.String(), HostNodePortCTZone))
 	} else {
 		// if externalIPOrLBIngressIP is not set then this service is of type NP
-		nodeportFlowsResult = append(nodeportFlowsResult,
-			// Traffic destined for ETP=local backed by OVN-K pod in LGW mode is matched on the svcNP and sent to table6.
-			fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=ct(commit,zone=%d,table=6)",
-				cookie, npw.ofportPhys, flowProtocol, svcPort, HostNodePortCTZone),
-			// Return traffic is matched on the svcNP as the sourcePort and gets un-DNATed back to nodeIP.
-			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, tp_src=%d, actions=ct(zone=%d,table=7)",
-				cookie, flowProtocol, svcPort, HostNodePortCTZone))
-	}
-	nodeportFlowsResult = append(nodeportFlowsResult,
+		// table 0, This rule matches on all traffic with dst port == NodePort, DNAT's it to the correct NodeIP
+		if strings.Contains(flowProtocol, "6") {
+			// If ipv6 make sure to choose the ipv6 node address for rule
+			nodeportFlows = append(nodeportFlows,
+				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=[%s]:%s),table=6)",
+					cookie, npw.ofportPhys, flowProtocol, svcPort.NodePort, HostNodePortCTZone, npw.gatewayIPv6, svcPort.TargetPort.String()))
+		} else {
+			nodeportFlows = append(nodeportFlows,
+				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=%s:%s),table=6)",
```

Comment:
we are missing comments on what these flows are doing

Reply:
its on L103, let me move it down.